File: src/main/java/work/lclpnet/ap2/game/guess_it/util/OptionMaker.java
package work.lclpnet.ap2.game.guess_it.util;

import it.unimi.dsi.fastutil.ints.IntArraySet;
import it.unimi.dsi.fastutil.ints.IntSet;
import net.minecraft.util.collection.IndexedIterable;

import java.util.List;
import java.util.Random;
import java.util.Set;
import java.util.function.IntFunction;

public class OptionMaker {

    private OptionMaker() {}

    public static <T> List<T> createOptions(Set<T> pool, int optionCount, Random random) {
        var index = List.copyOf(pool);
        return createOptions(index.size(), optionCount, random, index::get);
    }

    public static <T> List<T> createOptions(IndexedIterable<T> pool, int optionCount, Random random) {
        return createOptions(pool.size(), optionCount, random, pool::getOrThrow);
    }

    private static <T> List<T> createOptions(int poolSize, int optionCount, Random random, IntFunction<T> indexFunction) {
        if (poolSize < optionCount) {
            throw new IllegalStateException("There are less options available than requested");
        }

        IntSet indices = new IntArraySet(optionCount);

        while (indices.size() < optionCount) {
            int i = random.nextInt(poolSize);
            indices.add(i);
        }

        return indices.intStream()
                .mapToObj(indexFunction)
                .toList();
    }
}

File: src/minigames/guess_it/src/main/kotlin/work/lclpnet/ap2/game/guess_it/util/OptionMaker.kt
package work.lclpnet.ap2.game.guess_it.util

import net.minecraft.core.IdMap
import java.util.*

object OptionMaker {
    fun <T : Any> createOptions(pool: Set<T>, optionCount: Int, random: Random): List<T> {
        return createOptions(pool.toList(), optionCount, random)
    }

    fun <T : Any> createOptions(pool: List<T>, optionCount: Int, random: Random): List<T> {
        return createOptions(pool.size, optionCount, random) { index -> pool.get(index) }
    }

    fun <T : Any> createOptions(pool: IdMap<T>, optionCount: Int, random: Random): List<T> {
        return createOptions(pool.size(), optionCount, random) { id -> pool.byIdOrThrow(id) }
    }

    private fun <T> createOptions(
        poolSize: Int,
        optionCount: Int,
        random: Random,
        indexFunction: (Int) -> T
    ): List<T> {
        check(poolSize >= optionCount) { "There are less options available than requested" }

        val indices = mutableSetOf<Int>()

        while (indices.size < optionCount) {
            val i = random.nextInt(poolSize)
            indices.add(i)
        }

        return indices.map(indexFunction).toList()
    }
}
